x-pack/plugins/cases/public/common/hooks.ts

```diff
@@ -5,11 +5,120 @@
 import { STACK_APP_ID } from '../../common/constants';
 import { useCasesContext } from '../components/cases_context/use_cases_context';
 
+import { DEFAULT_QUERY_PARAMS } from '../containers/use_get_cases';
+
+import type {
+  LocalStorageQueryParams,
+  QueryParams,
+  ParsedUrlQueryParams,
+  UrlQueryParams,
+} from '../../common/ui/types';
+
+const parseUrlQueryParams = (search: string): UrlQueryParams => {
+  const parsedUrlParams: ParsedUrlQueryParams = parse(search);
+
+  const urlParams: UrlQueryParams = {
+    ...(parsedUrlParams.sortField && { sortField: parsedUrlParams.sortField }),
+    ...(parsedUrlParams.sortOrder && { sortOrder: parsedUrlParams.sortOrder }),
+  };
+
+  const intPage = parsedUrlParams.page && parseInt(parsedUrlParams.page, 10);
+  const intPerPage = parsedUrlParams.perPage && parseInt(parsedUrlParams.perPage, 10);
+
+  if (intPage) {
+    urlParams.page = intPage;
+  }
+
+  if (intPerPage) {
+    urlParams.perPage = intPerPage;
+  }
+
+  return urlParams;
+};
+
 export const useIsMainApplication = () => {
   const { appId } = useCasesContext();
 
   return appId === STACK_APP_ID;
 };
+
+export function useUrlState(isSelectorView: boolean = false) {
+  const location = useLocation();
+  const history = useHistory();
+
+  const [queryParams, setQueryParams] = useState<UrlQueryParams>({});
+
+  const [localStorageQueryParams, setLocalStorageQueryParams] =
+    useLocalStorage<LocalStorageQueryParams>('cases.list.preferences');
+
+  const _setUrlQueryParams = useCallback(
+    (params) => {
+      const urlParams: UrlQueryParams = parseUrlQueryParams(location.search);
+      const newQueryParams: QueryParams = { ...DEFAULT_QUERY_PARAMS };
+
+      newQueryParams.perPage =
+        params.perPage ??
+        urlParams.perPage ??
+        localStorageQueryParams?.perPage ??
+        DEFAULT_QUERY_PARAMS.perPage;
+      newQueryParams.sortField = params.sortField ?? DEFAULT_QUERY_PARAMS.sortField;
+      newQueryParams.sortOrder =
+        params.sortOrder ??
+        urlParams.sortOrder ??
+        localStorageQueryParams?.sortOrder ??
+        DEFAULT_QUERY_PARAMS.sortOrder;
+      newQueryParams.page = params.page ?? urlParams.page ?? DEFAULT_QUERY_PARAMS.page;
+
+      if (stringify(queryParams) !== stringify(newQueryParams)) {
+        const newLocalStorageQueryParams = {
+          perPage: newQueryParams.perPage,
+          sortOrder: newQueryParams.sortOrder,
+        };
+
+        setLocalStorageQueryParams(newLocalStorageQueryParams);
+
+        const newUrlParams = {
+          page: newQueryParams.page,
+          ...newLocalStorageQueryParams,
+        };
+
+        history.replace({
+          ...location,
+          search: stringify(newUrlParams),
+        });
+
+        setQueryParams(newQueryParams);
+      }
+    },
+    [history, localStorageQueryParams, location, setLocalStorageQueryParams, queryParams]
+  );
+
+  const setUrlQueryParams = useCallback(
+    (params) => {
+      if (isSelectorView) {
```

Comment:
Just curious what is the selector view? Is that when you're doing bulk actions?

Reply:
If `isSelectorView` is true, the component is rendered inside a modal(where we don't want the new logic to apply).

Reply:
Ah ok, I know it's called `isSelectorView` outside of this function but how about we rename it to something like `isModalView`. That way I don't forget and have to ask again 😆